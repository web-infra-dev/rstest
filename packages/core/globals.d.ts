declare global {
  const test: typeof import('@rstest/core')['test'];
  const describe: typeof import('@rstest/core')['describe'];
  const it: typeof import('@rstest/core')['it'];
  const expect: typeof import('@rstest/core')['expect'];
}
export {};
