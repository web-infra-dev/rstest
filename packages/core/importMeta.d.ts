interface ImportMeta {
  readonly rstest?: typeof import('@rstest/core');
}
