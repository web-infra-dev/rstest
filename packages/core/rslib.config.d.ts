import { RslibConfig } from '@rslib/core';
declare const config: RslibConfig;
export default config;
