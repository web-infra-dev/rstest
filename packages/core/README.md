<picture>
  <img alt="Rstest Banner" src="https://assets.rspack.dev/rstest/rstest-banner.png">
</picture>

# Rstest

Rstest is a testing framework powered by Rspack.

## ðŸ“– License

Rstest is licensed under the [MIT License](https://github.com/web-infra-dev/rstest/blob/main/LICENSE).
