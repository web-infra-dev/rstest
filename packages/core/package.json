{
  "name": "@rstest/core",
  "version": "0.0.0",
  "description": "The Rsbuild-based test tool.",
  "bugs": {
    "url": "https://github.com/web-infra-dev/rstest/issues"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/web-infra-dev/rstest",
    "directory": "packages/core"
  },
  "license": "MIT",
  "type": "module",
  "main": "./dist/index.js",
  "types": "./dist-types/index.d.ts",
  "bin": {
    "rstest": "./bin/rstest.js"
  },
  "exports": {
    ".": {
      "types": "./dist-types/index.d.ts",
      "default": "./dist/index.js"
    },
    "./globals": {
      "types": "./globals.d.ts"
    }
  },
  "files": [
    "bin",
    "dist",
    "dist-types",
    "compiled",
    "types.d.ts"
  ],
  "scripts": {
    "build": "rslib build",
    "dev": "rslib build --watch",
    "test": "npx rstest run --globals"
  },
  "dependencies": {
    "@rsbuild/core": "^1.2.19",
    "birpc": "2.2.0",
    "chai": "^5.2.0",
    "std-env": "^3.8.1",
    "tinypool": "^1.0.2"
  },
  "devDependencies": {
    "@babel/code-frame": "^7.26.2",
    "@jridgewell/trace-mapping": "0.3.25",
    "@rslib/core": "0.5.5",
    "@rstest/tsconfig": "workspace:*",
    "@types/babel__code-frame": "^7.0.6",
    "@vitest/expect": "^3.0.9",
    "cac": "^6.7.14",
    "picocolors": "^1.1.1",
    "rslog": "^1.2.3",
    "stacktrace-parser": "0.1.11",
    "tinyglobby": "^0.2.12"
  },
  "engines": {
    "node": ">=16.7.0"
  },
  "publishConfig": {
    "access": "public",
    "provenance": true,
    "registry": "https://registry.npmjs.org/"
  }
}
