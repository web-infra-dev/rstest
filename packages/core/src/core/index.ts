import type { RstestCommand, RstestConfig, RstestInstance } from '../types';
import { createContext } from './context';

export function createRstest(
  config: RstestConfig,
  command: RstestCommand,
  fileFilters: string[],
): RstestInstance {
  const context = createContext({ cwd: process.cwd(), command }, config);

  const runTests = async (): Promise<void> => {
    const { runTests } = await import('./runTests');
    await runTests(context, fileFilters);
  };

  return {
    runTests,
  };
}
