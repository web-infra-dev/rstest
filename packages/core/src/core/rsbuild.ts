import fs from 'node:fs';
import {
  type RsbuildInstance,
  logger as RsbuildLogger,
  type RsbuildPlugin,
  type Rspack,
  createRsbuild,
} from '@rsbuild/core';
import path from 'pathe';
import type { EntryInfo, SourceMapInput } from '../types';
import { isDebug } from '../utils';
import { pluginIgnoreResolveError } from './plugins/ignoreResolveError';

const isMultiCompiler = <
  C extends Rspack.Compiler = Rspack.Compiler,
  M extends Rspack.MultiCompiler = Rspack.MultiCompiler,
>(
  compiler: C | M,
): compiler is M => {
  return 'compilers' in compiler && Array.isArray(compiler.compilers);
};

class TestFileWatchPlugin {
  private contextToWatch: string | null = null;

  constructor(contextToWatch: string) {
    this.contextToWatch = contextToWatch;
  }

  apply(compiler: Rspack.Compiler) {
    compiler.hooks.afterCompile.tap(
      'Rstest:TestFileWatchPlugin',
      (compilation) => {
        if (this.contextToWatch === null) {
          return;
        }

        const contextDep = compilation.contextDependencies;
        if (!contextDep.has(this.contextToWatch)) {
          contextDep.add(this.contextToWatch);
        }
      },
    );
  }
}

export const prepareRsbuild = async (
  name: string,
  globTestSourceEntries: () => Promise<Record<string, string>>,
  setupFiles: Record<string, string>,
): Promise<RsbuildInstance> => {
  RsbuildLogger.level = isDebug() ? 'verbose' : 'error';

  const rsbuildInstance = await createRsbuild({
    rsbuildConfig: {
      server: {
        printUrls: false,
        strictPort: false,
        middlewareMode: true,
      },
      environments: {
        [name]: {
          dev: {
            writeToDisk: false,
          },
          output: {
            sourceMap: {
              js: 'source-map',
            },
            externals: {
              '@rstest/core': 'global @rstest/core',
            },
            target: 'node',
          },
          tools: {
            rspack: (config) => {
              config.output ??= {};
              config.output.devtoolModuleFilenameTemplate =
                '[absolute-resource-path]';

              config.optimization = {
                ...(config.optimization || {}),
                moduleIds: 'named',
                chunkIds: 'named',
              };

              config.plugins!.push(new TestFileWatchPlugin(process.cwd()));
              config.entry = async () => {
                const sourceEntries = await globTestSourceEntries();
                return {
                  ...sourceEntries,
                  ...setupFiles,
                };
              };
            },
          },
          plugins: [pluginIgnoreResolveError],
        },
      },
    },
  });

  return rsbuildInstance;
};

export const createRsbuildServer = async ({
  name,
  globTestSourceEntries,
  setupFiles,
  rsbuildInstance,
  rootPath,
}: {
  rsbuildInstance: RsbuildInstance;
  name: string;
  globTestSourceEntries: () => Promise<Record<string, string>>;
  setupFiles: Record<string, string>;
  rootPath: string;
}): Promise<
  () => Promise<{
    buildTime: number;
    entries: EntryInfo[];
    setupEntries: EntryInfo[];
    assetFiles: Record<string, string>;
    sourceMaps: Record<string, SourceMapInput>;
    getSourcemap: (sourcePath: string) => SourceMapInput | null;
    close: () => Promise<void>;
  }>
> => {
  // Read files from memory via `rspackCompiler.outputFileSystem`
  let rspackCompiler: Rspack.Compiler | Rspack.MultiCompiler;

  const rstestCompilerPlugin: RsbuildPlugin = {
    name: 'rstest:compiler',
    setup: (api) => {
      api.onAfterCreateCompiler(({ compiler }) => {
        // outputFileSystem to be updated later by `rsbuild-dev-middleware`
        rspackCompiler = compiler;
      });

      api.modifyRspackConfig((config) => {
        config?.plugins?.push(new TestFileWatchPlugin(rootPath));
      });
    },
  };

  rsbuildInstance.addPlugins([rstestCompilerPlugin]);

  const devServer = await rsbuildInstance.createDevServer({
    getPortSilently: true,
  });

  if (isDebug()) {
    await rsbuildInstance.inspectConfig({ writeToDisk: true });
  }

  const outputFileSystem =
    (isMultiCompiler(rspackCompiler!)
      ? rspackCompiler.compilers[0]!.outputFileSystem
      : rspackCompiler!.outputFileSystem) || fs;

  const getRsbuildStats = async () => {
    const stats = await devServer.environments[name]!.getStats();

    const {
      entrypoints,
      outputPath,
      assets,
      time: buildTime,
    } = stats.toJson({
      entrypoints: true,
      outputPath: true,
      assets: true,
      // get the compilation time
      timings: true,
    });

    const readFile = async (fileName: string) => {
      return new Promise<string>((resolve, reject) => {
        outputFileSystem.readFile(fileName, (err, data) => {
          if (err) {
            reject(err);
          }
          resolve(typeof data === 'string' ? data : data!.toString());
        });
      });
    };

    const entries: EntryInfo[] = [];
    const setupEntries: EntryInfo[] = [];
    const sourceEntries = await globTestSourceEntries();
    // TODO: check compile error, such as setupFiles not found
    for (const entry of Object.keys(entrypoints!)) {
      const e = entrypoints![entry]!;

      const filePath = path.join(
        outputPath!,
        e.assets![e.assets!.length - 1]!.name,
      );

      if (setupFiles[entry]) {
        setupEntries.push({
          filePath,
          originPath: setupFiles[entry],
        });
      } else if (sourceEntries[entry]) {
        entries.push({
          filePath,
          originPath: sourceEntries[entry],
        });
      }
    }

    const sourceMaps: Record<string, SourceMapInput> = Object.fromEntries(
      (
        await Promise.all(
          assets!.map(async (asset) => {
            const sourceMapPath = asset?.info.related?.sourceMap?.[0];

            const assetFilePath = path.join(outputPath!, asset.name);
            if (sourceMapPath) {
              const filePath = path.join(outputPath!, sourceMapPath);
              const sourceMap = await readFile(filePath);
              return [assetFilePath, JSON.parse(sourceMap)];
            }
            return [assetFilePath, null];
          }),
        )
      ).filter((asset) => asset[1] !== null),
    );

    return {
      entries,
      setupEntries,
      buildTime: buildTime!,
      // Resources need to be obtained synchronously when the test is loaded, so files need to be read in advance
      assetFiles: Object.fromEntries(
        await Promise.all(
          assets!.map(async (a) => {
            const filePath = path.join(outputPath!, a.name);
            return [filePath, await readFile(filePath)];
          }),
        ),
      ),
      sourceMaps,
      getSourcemap: (sourcePath: string): SourceMapInput | null => {
        return sourceMaps[sourcePath] || null;
      },
      close: devServer.close,
    };
  };

  return getRsbuildStats;
};
