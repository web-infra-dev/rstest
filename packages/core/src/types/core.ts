import type { SnapshotManager } from '@vitest/snapshot/manager';
import type { NormalizedConfig, RstestConfig } from './config';
import type { Reporter } from './reporter';

export type RstestCommand = 'watch' | 'run';

export type RstestContext = {
  /** The Rstest core version. */
  version: string;
  /** The root path of current project. */
  rootPath: string;
  /** The original Rstest config passed from the createRstest method. */
  originalConfig: Readonly<RstestConfig>;
  /** The normalized Rstest config. */
  normalizedConfig: NormalizedConfig;
  /**
   * The command type.
   *
   * - dev: `rstest dev`
   * - run: `rstest run`
   */
  command: RstestCommand;
  reporters: Reporter[];
  snapshotManager: SnapshotManager;
};

export type RstestInstance = {
  runTests: () => Promise<void>;
};
