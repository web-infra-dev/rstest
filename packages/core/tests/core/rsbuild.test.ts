import { prepareRsbuild } from '../../src/core/rsbuild';
import type { RstestContext } from '../../src/types';

describe('prepareRsbuild', () => {
  it('should generate rspack config correctly', async () => {
    const rsbuildInstance = await prepareRsbuild(
      {
        normalizedConfig: {
          name: 'test',
          plugins: [],
          resolve: {},
          source: {},
          output: {},
          tools: {},
        },
      } as unknown as RstestContext,
      async () => ({}),
      {},
    );
    expect(rsbuildInstance).toBeDefined();
    const {
      origin: { bundlerConfigs },
    } = await rsbuildInstance.inspectConfig();

    expect(bundlerConfigs[0]).toMatchSnapshot();
  });

  it('should generate swc config correctly with user customize', async () => {
    const rsbuildInstance = await prepareRsbuild(
      {
        normalizedConfig: {
          name: 'test',
          plugins: [],
          resolve: {},
          source: {
            decorators: {
              version: 'legacy',
            },
            include: [/node_modules[\\/]query-string[\\/]/],
          },
          output: {},
          tools: {},
        },
      } as unknown as RstestContext,
      async () => ({}),
      {},
    );
    expect(rsbuildInstance).toBeDefined();
    const {
      origin: { bundlerConfigs },
    } = await rsbuildInstance.inspectConfig();

    expect(
      bundlerConfigs[0]?.module?.rules?.filter(
        (rule) =>
          rule &&
          typeof rule === 'object' &&
          rule.test &&
          rule.test instanceof RegExp &&
          rule.test.test('a.js'),
      ),
    ).toMatchSnapshot();
  });
});
