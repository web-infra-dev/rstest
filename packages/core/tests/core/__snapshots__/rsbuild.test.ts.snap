// Snapshot v1

exports[`prepareRsbuild > should generate rspack config correctly 1`] = `
{
  "context": "<WORKSPACE>",
  "devtool": "source-map",
  "entry": {},
  "experiments": {
    "asyncWebAssembly": true,
    "incremental": true,
  },
  "externals": [
    {
      "@rstest/core": "global @rstest/core",
    },
    [Function],
  ],
  "externalsPresets": {
    "node": true,
  },
  "ignoreWarnings": [
    /Module not found/,
  ],
  "infrastructureLogging": {
    "level": "error",
  },
  "mode": "none",
  "module": {
    "parser": {
      "javascript": {
        "exportsPresence": "error",
        "importDynamic": false,
      },
    },
    "rules": [
      {
        "resolve": {
          "fullySpecified": false,
        },
        "test": /\\\\\\.m\\?js/,
      },
      {
        "dependency": {
          "not": "url",
        },
        "resolve": {
          "preferRelative": true,
        },
        "resourceQuery": {
          "not": /raw\\|inline/,
        },
        "sideEffects": true,
        "test": /\\\\\\.css\\$/,
        "type": "javascript/auto",
        "use": [
          {
            "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rsbuild/core/dist/ignoreCssLoader.mjs",
          },
          {
            "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rsbuild/core/compiled/css-loader/index.js",
            "options": {
              "importLoaders": 0,
              "modules": {
                "auto": true,
                "exportGlobals": false,
                "exportLocalsConvention": "camelCase",
                "exportOnlyLocals": true,
                "localIdentName": "[path][name]__[local]-[hash:base64:6]",
                "namedExport": false,
              },
              "sourceMap": false,
            },
          },
        ],
      },
      {
        "resolve": {
          "preferRelative": true,
        },
        "resourceQuery": /inline/,
        "sideEffects": true,
        "test": /\\\\\\.css\\$/,
        "type": "javascript/auto",
        "use": [
          {
            "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rsbuild/core/compiled/css-loader/index.js",
            "options": {
              "exportType": "string",
              "importLoaders": 0,
              "modules": false,
              "sourceMap": false,
            },
          },
        ],
      },
      {
        "resourceQuery": /raw/,
        "test": /\\\\\\.css\\$/,
        "type": "asset/source",
      },
      {
        "dependency": {
          "not": "url",
        },
        "include": [
          {
            "and": [
              "<WORKSPACE>",
              {
                "not": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
              },
            ],
          },
          /\\\\\\.\\(\\?:ts\\|tsx\\|jsx\\|mts\\|cts\\)\\$/,
        ],
        "test": /\\\\\\.\\(\\?:js\\|jsx\\|mjs\\|cjs\\|ts\\|tsx\\|mts\\|cts\\)\\$/,
        "type": "javascript/auto",
        "use": [
          {
            "loader": "builtin:swc-loader",
            "options": {
              "env": {
                "targets": [
                  "node >= 16",
                ],
              },
              "isModule": "unknown",
              "jsc": {
                "experimental": {
                  "cacheRoot": "<WORKSPACE>/node_modules/.cache/.swc",
                  "keepImportAttributes": true,
                },
                "externalHelpers": true,
                "parser": {
                  "decorators": true,
                  "syntax": "typescript",
                  "tsx": false,
                },
                "transform": {
                  "decoratorVersion": "2022-03",
                  "legacyDecorator": false,
                },
              },
            },
          },
        ],
      },
      {
        "mimetype": {
          "or": [
            "text/javascript",
            "application/javascript",
          ],
        },
        "resolve": {
          "fullySpecified": false,
        },
        "use": [
          {
            "loader": "builtin:swc-loader",
            "options": {
              "env": {
                "targets": [
                  "node >= 16",
                ],
              },
              "isModule": "unknown",
              "jsc": {
                "experimental": {
                  "cacheRoot": "<WORKSPACE>/node_modules/.cache/.swc",
                  "keepImportAttributes": true,
                },
                "externalHelpers": true,
                "parser": {
                  "decorators": true,
                  "syntax": "typescript",
                  "tsx": false,
                },
                "transform": {
                  "decoratorVersion": "2022-03",
                  "legacyDecorator": false,
                },
              },
            },
          },
        ],
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/image/[name].[contenthash:8][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "resourceQuery": /raw/,
            "type": "asset/source",
          },
          {
            "generator": {
              "filename": "static/image/[name].[contenthash:8][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 4096,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:png\\|jpg\\|jpeg\\|pjpeg\\|pjp\\|gif\\|bmp\\|webp\\|ico\\|apng\\|avif\\|tif\\|tiff\\|jfif\\|cur\\)\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/svg/[name].[contenthash:8].svg",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "resourceQuery": /raw/,
            "type": "asset/source",
          },
          {
            "generator": {
              "filename": "static/svg/[name].[contenthash:8].svg",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 4096,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.svg\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/media/[name].[contenthash:8][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "resourceQuery": /raw/,
            "type": "asset/source",
          },
          {
            "generator": {
              "filename": "static/media/[name].[contenthash:8][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 4096,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:mp4\\|webm\\|ogg\\|mov\\|mp3\\|wav\\|flac\\|aac\\|m4a\\|opus\\)\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/font/[name].[contenthash:8][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "resourceQuery": /raw/,
            "type": "asset/source",
          },
          {
            "generator": {
              "filename": "static/font/[name].[contenthash:8][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 4096,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:woff\\|woff2\\|eot\\|ttf\\|otf\\|ttc\\)\\$/i,
      },
      {
        "dependency": "url",
        "generator": {
          "filename": "static/wasm/[hash].module.wasm",
        },
        "test": /\\\\\\.wasm\\$/,
        "type": "asset/resource",
      },
      {
        "test": /\\\\\\.node\\$/,
        "use": [
          {
            "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/@rsbuild/core/dist/transformRawLoader.mjs",
            "options": {
              "getEnvironment": [Function],
              "id": "rsbuild-transform-0",
            },
          },
        ],
      },
    ],
  },
  "name": "test",
  "optimization": {
    "chunkIds": "named",
    "emitOnErrors": true,
    "minimize": false,
    "moduleIds": "named",
    "splitChunks": false,
  },
  "output": {
    "assetModuleFilename": "static/assets/[name].[contenthash:8][ext]",
    "chunkFilename": "[name].js",
    "devtoolModuleFilenameTemplate": "[absolute-resource-path]",
    "filename": "[name].js",
    "hashFunction": "xxhash64",
    "iife": false,
    "library": {
      "type": "commonjs2",
    },
    "path": "<WORKSPACE>/dist/.test",
    "pathinfo": false,
    "publicPath": "/",
    "webassemblyModuleFilename": "static/wasm/[hash].module.wasm",
  },
  "performance": {
    "hints": false,
  },
  "plugins": [
    {
      "name": "RsbuildCorePlugin",
    },
    DefinePlugin {
      "_args": [
        {
          "import.meta.dirname": "__dirname",
          "import.meta.env.ASSET_PREFIX": "\\"\\"",
          "import.meta.env.BASE_URL": "\\"/\\"",
          "import.meta.env.DEV": false,
          "import.meta.env.MODE": "\\"none\\"",
          "import.meta.env.PROD": false,
          "import.meta.filename": "__filename",
          "import.meta.rstest": "global['@rstest/core']",
          "process.env.ASSET_PREFIX": "\\"\\"",
          "process.env.BASE_URL": "\\"/\\"",
        },
      ],
      "affectedHooks": "compilation",
      "name": "DefinePlugin",
    },
    IgnoreModuleNotFoundErrorPlugin {},
  ],
  "resolve": {
    "alias": {
      "@swc/helpers": "<WORKSPACE>/node_modules/<PNPM_INNER>/@swc/helpers",
    },
    "extensions": [
      ".ts",
      ".tsx",
      ".mjs",
      ".js",
      ".jsx",
      ".json",
    ],
  },
  "target": "node",
  "watchOptions": {
    "aggregateTimeout": 0,
  },
}
`;

exports[`prepareRsbuild > should generate swc config correctly with user customize 1`] = `
[
  {
    "resolve": {
      "fullySpecified": false,
    },
    "test": /\\\\\\.m\\?js/,
  },
  {
    "dependency": {
      "not": "url",
    },
    "include": [
      {
        "and": [
          "<WORKSPACE>",
          {
            "not": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
          },
        ],
      },
      /\\\\\\.\\(\\?:ts\\|tsx\\|jsx\\|mts\\|cts\\)\\$/,
      /node_modules\\[\\\\\\\\/\\]query-string\\[\\\\\\\\/\\]/,
    ],
    "test": /\\\\\\.\\(\\?:js\\|jsx\\|mjs\\|cjs\\|ts\\|tsx\\|mts\\|cts\\)\\$/,
    "type": "javascript/auto",
    "use": [
      {
        "loader": "builtin:swc-loader",
        "options": {
          "env": {
            "targets": [
              "node >= 16",
            ],
          },
          "isModule": "unknown",
          "jsc": {
            "experimental": {
              "cacheRoot": "<WORKSPACE>/node_modules/.cache/.swc",
              "keepImportAttributes": true,
            },
            "externalHelpers": true,
            "parser": {
              "decorators": true,
              "syntax": "typescript",
              "tsx": false,
            },
            "transform": {
              "decoratorMetadata": true,
              "legacyDecorator": true,
              "useDefineForClassFields": false,
            },
          },
        },
      },
    ],
  },
]
`;
